README.md
# playercenter-backend

A backend to gather and store data about players.

## API

### Retrieve public information about a player

Does not need any authentication. The response from this endpoint is very close to the non-public information about players (**GET** ``/:UUID:``) . It omits any sensible data (e.g. venue specific IDs), though.

#### Request

**GET** ``/public/:UUID:``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player you want to retrieve information about.

##### Body

Empty.

#### Response

##### Body

JSON encoded object like this:

```javascript
{
  "uuid": "some-uuid",
  "venues": {
    "facebook": {
      "name": "The Peter"
    },
    "spiral-galaxy": {
      "name": "Peter Smith"
    }
  }
}
```

All information under the ``venues`` is present for each venue this player is playing a Quarter Spiral game on.

### Retrieve public information about a bunch of players

Same as retrieving public information about a single player but for multiple ones at once.

#### Request

**GET** ``/public/players``

##### Parameters

- **uuids** [REQUIRED]: A JSON encoded array of UUIDs

##### Body

Empty.

#### Response

##### Body


JSON encoded object like this:

```javascript
{
  "some-uuid": {
    "uuid": "some-uuid",
    "venues": {
      "facebook": {
        "name": "The Peter"
      },
      "spiral-galaxy": {
        "name": "Peter Smith"
      }
    }
  },
  "another-uuid": {
    "uuid": "another-uuid",
    "venues": {
      "facebook": {
        "name": "The Hans"
      },
      "spiral-galaxy": {
        "name": "Hans Franz"
      }
    }
  },
}
```

### List player's friends public information

This endpoint does not need any authentication and resembles closely the ``List player's friends`` endpoint (**GET** to ``/:UUID:/friends``) which needs to be called authenticated over OAuth. This endpoint will only return publicly available information about friends.

#### Request

**GET** ``/public/:UUID:/friends``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player who's friends you want to retrieve.

##### Body

Empty

#### Response

Please note that the list of friends always includes the requester itself, too.

##### Body

JSON encoded object mapping a friend's UUID to an object of player info (see *Retrieve public information about a player*) like this:

```javascript
{
  "some-uuid":  {
    "uuid": "some-uuid",
    "facebook": {
      "name": "The Peter"
    },
    "spiral-galaxy": {
      "name": "Peter Smith"
    }
  },
  "other-uuid": {
    …
  }
}
```

### List the games a player plays

This endpoint does not need authentication.

#### Request

**GET** to ``/public/:UUID:/games``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player who's games are going to be retrieved

##### Body

JSON encoded Object with additional options:

* **venue**: If set only games that the player play's on the given venue will be listed

#### Response

##### Body

JSON encoded object of games in the way the ``devcenter-backend`` public API returns game lists.

### Retrieve information about a player

#### Request

**GET** ``/:UUID:``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player you want to retrieve information about.

##### Body

Empty.

#### Response

##### Body

JSON encoded object like this:

```javascript
{
  "uuid": "some-uuid",
  "venues": {
    "facebook": {
      "id": "1234",
      "name": "The Peter"
    },
    "spiral-galaxy": {
      "id": "87233",
      "name": "Peter Smith"
    }
  }
}
```

All information under the ``venues`` is present for each venue this player is playing a Quarter Spiral game on.

### List player's friends

#### Request

**GET** ``/:UUID:/friends``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player who's friends you want to retrieve.

##### Body

JSON encoded object with further options:

* **game**: If present only displays friends who play the game with the given UUID
* **meta**: An array of keys of player meta data that is added to each friend. This parameter requires the game parameter to be set as the meta data is stored per player/game pair.

#### Response

Please note that the list of friends always includes the requester itself, too.

##### Body

JSON encoded object mapping a friend's UUID to an object of player info (see *Retrieve information about a player*) like this:

```javascript
{
  "some-uuid":  {
    "uuid": "some-uuid",
    "facebook": {
      "id": "1234",
      "name": "The Peter"
    },
    "spiral-galaxy": {
      "id": "87233",
      "name": "Peter Smith"
    }
  },
  "other-uuid": {
    …
  }
}
```

If the ``meta`` parameter is set the meta results will be added as another venue called ``meta`` like this:

```javascript
{
  "some-uuid":  {
    "uuid": "some-uuid",
    "facebook": {
      "id": "1234",
      "name": "The Peter"
    },
    "spiral-galaxy": {
      "id": "87233",
      "name": "Peter Smith"
    },
    "meta": {
      "highScore": 100,
      "lastLevel": "Good Chamber"
    }
  },
  "other-uuid": {
    …
  }
}
```

### [DEPRECATED] List the games a player plays

#### Request

**GET** to ``/:UUID:/games``

This endpoint is deprecated in favor of the public endpoint at ``/public/:UUID:/games`` which behaves the same but does not require authentication. You must use the public endpoint instead.

### List the games the friends of a player play

#### Request

**GET** to ``/:UUID:/games/friends``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player who's games are going to be retrieved

##### Body

JSON encoded Object with additional options:

* **venue**: If set only games that the player play's on the given venue will be listed

#### Response

##### Body

JSON encoded object of games in the way the ``devcenter-backend`` public API returns game lists.

### Update a users friends

#### Request

**PUT** to ``/:UUID:/friends/:VENUE:``

This will add all friends that are not already in the system.

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player who's friends are going to be updated
- **VENUE** [REQUIRED]: The venue on which the friendships are etablished

##### Body

A JSON encoded object of venue identities like this:

```javascript
{
  "friends": [
    {"venue-id": "23423", "name": "Peter Smith"},
    {"venue-id": "89482", "name": "Sam Jackson"}
  ]
}
```

#### Response

Returns status 200 on success.

##### Body

Empty.

### Retrieve a user's avatar on a given venue

This request does not require any authentication.

#### Request

**GET** to ``/:UUID:/avatars/:VENUE:``

##### Parameters

- **UUID** [REQUIRED]: The UUID of the player who's avatar is going to be retrieved
- **VENUE** [REQUIRED]: The venue of which the user's avatar is going to be retrieved from

##### Body

Empty.

#### Response

This might return the avatar image directly (HTTP status ``200``) or a ``302`` HTTP redirect to the actual avatar image.

### Register a user as a player of a game

#### Request

**POST** to ``/:PLAYER-UUID:/games/:GAME-UUID:/:VENUE:``

##### Parameters

- **PLAYER-UUID** [REQUIRED]: The UUID of the player who is playing the game
- **GAME-UUID** [REQUIRED]: The UUID of a game the player is playing
- **VENUE** [REQUIRED]: The venue on which the player is playing

##### Body

Empty.

#### Response

Returns status code 201 if the player was registered as a new player of the game, 304 if the player already plays the game on that venue. If the player already plays the game but not on that venue a 200 status code is returned instead. The body is empty.


### Set player's game meta data

#### Request

**PUT** to ``/:PLAYER-UUID:/games/:GAME-UUID:/meta-data/:KEY:``

##### Parameters

- **PLAYER-UUID** [REQUIRED]: The UUID of the player who you want to set the meta data for
- **GAME-UUID** [REQUIRED]: The UUID of a game the you want to set the meta data for
- **KEY** [OPTIONAL] A key name in the metadata object. If set only this key will be updated.

##### Body

JSON encoded object of meta data under a ``meta`` key like this:

```javascript
{
  "meta": {
    "highscore": 100,
    "levelsCompleted": "[1,2,3]",
    "playedTutorial": true
  }
}
```

Only numbers, boolean values and strings are accepted as meta data values. If you need more complex types as arrays and objects you can always encode them as strings, e.g. in JSON notation.

If the ``KEY`` parameter is set you still provide an object as the body. Nonetheless, the request will only change the specified key of the metadata. E.g. if you only want to change the high score to ``112``, you can issue a **PUT** to ``/:PLAYER-UUID:/games/:GAME-UUID:/meta-data/highscore`` with a JSON encoded body of this object:

```javascript
{
  "meta": {
    "highscore": 110
  }
}
```

#### Response

Returns status code 200 if the data was set successfully. Returns status code 415 if wrong data types are used in the meta data.

##### Body

JSON encoded meta data object of the changed meta data. This always returns the whole meta data object even when a ``KEY`` parameter was present.

Example:

```javascript
{
  "meta": {
    "highscore": 110,
    "levelsCompleted": "[1,2,3]",
    "playedTutorial": true
  }
}
```

### Retrieve player's game meta data

#### Request

**GET** to ``/:PLAYER-UUID:/games/:GAME-UUID:/meta-data``

##### Parameters

- **PLAYER-UUID** [REQUIRED]: The UUID of the player who you want to get the meta data for
- **GAME-UUID** [REQUIRED]: The UUID of a game the you want to set get meta data for

##### Body

Empty.

#### Response

JSON encoded meta data object.

Example:

```javascript
{
  "meta": {
    "highscore": 110,
    "levelsCompleted": "[1,2,3]",
    "playedTutorial": true
  }
}
```

CHANGELOG.md
# 0.0.39 / 2013-08-12

* Fixed sentry app id

# 0.0.38 / 2013-08-12

* Adds getsentry.com exception tracking
* Adds request id tracker
* Bumps client dependencies

# 0.0.37  / 2013-07-24

* Fills up venue identities automatically on game registration

# 0.0.36 / 2013-07-12

* Fix a bug when switching to futuro

# 0.0.35 / 2013-07-12

* Moves friend updates out of the request/response cycle to allow large friend updates without timeout

# 0.0.34 / 2013-07-08

* Fixes result format when requesting public venue information as a batch for a single player

# 0.0.33 / 2013-07-08

* Adds an endpoint to retrieve public venue identity information in batches

# 0.0.32 / 2013-05-23

* Makes it possible to fake HTTP methods through a header
* Makes it possible to pass the OAuth token as a query param

# 0.0.31 / 2013-05-14

* Adds a ``crossdomain.xml``

# 0.0.30 / 2013-04-13

* Updates friends only once per day for each user on each venue

# 0.0.29 / 2013-04-02

* Adds authorization

# 0.0.28

* Fixes a cached bad response format

# 0.0.27

* Updates dependencies

# 0.0.26

* Makes meta data retrieval work again for not yet set meta data

# 0.0.25

* Adds caching to own and friend's games list

# 0.0.24

* Adds public endpoint to retrieve games a user plays
* Adds public endpoint to retrieve a user's friends

# 0.0.23

* Adds an endpoint for public player information
* Fixes documentation errors

# 0.0.22

* Fixes CORS headers

# 0.0.21

* Includes the requester into all friends lists

# 0.0.20

* Improve Newrelic instrumenting

# 0.0.19

* Adds Newrelic monitoring and ping middleware

# 0.0.18

* Adds player meta data to friend retrieval end point

# 0.0.17

* Fixes a bug saving ``false`` values in the game/player meta data

# 0.0.16

* Adds game/player metadata endpoints

# 0.0.15

* Adds CORS headers and OPTION endpoints to the API

# 0.0.14

* Adds an endpoint to list all games a player plays
* Adds an endpoint to list friends games
* Adds Thin

# 0.0.13

* Adds an endpoint to make a user a player in the QS graph

# 0.0.12

* Adds an endpoint to retrieve a player's avatar picture on a given venue

# 0.0.11

* Makes the API CORS compliant

# 0.0.10

* Eases the grape dependency

# 0.0.9

* Refactors galaxy-spiral to spiral-galaxy in the README
* Bumps auth-backend and auth-client

# 0.0.8

* Bumps graph-client to handle empty response bodies from the graph-backend without errors

# 0.0.7

* Changes the content type of the API to be UTF8 encoded
* Makes it runnable on metaserver

# 0.0.6

- Loads graph-client to make it work on heroku

# 0.0.5

- Adds config.ru to really make this run on heroku

# 0.0.4

- Adds Gemfile.lock to git to make the app run on heroku

# 0.0.3

- Fixes specs to make them order agnostic
- Fixes the dependencies (pin auth-client, add graph-client)

# 0.0.2

- The beginning


lib/playercenter-backend/api.rb
require 'grape'
require 'grape_newrelic'

module Playercenter::Backend
  class API < ::Grape::API
    use GrapeNewrelic::Instrumenter

    class TokenStore
      def self.token(connection)
        @token ||= connection.auth.create_app_token(ENV['QS_OAUTH_CLIENT_ID'], ENV['QS_OAUTH_CLIENT_SECRET'])
      end

      def self.reset!
        @token = nil
      end
    end

    version 'v1', :using => :path, :vendor => 'quarter-spiral'

    format :json
    default_format :json

    default_error_formatter :json

    rescue_from Error::InvalidPlayerMetaDataError do |e|
        Rack::Response.new({
            'status' => 415,
            'message' => e.message
        }.to_json, 415)
    end

    helpers do
      def connection
        @connection ||= Connection.create
      end

      def not_found!
        error!('Not found', 404)
      end

      def token
        TokenStore.token(connection)
      end

      def try_twice_and_avoid_token_expiration
        yield
      rescue Service::Client::ServiceError => e
        raise e unless e.error == 'Unauthenticated'
        TokenStore.reset!
        yield
      end

      def authentication_exception?
        env['PATH_INFO'] =~ /\/avatars\/[^\/]+$/ ||
env['PATH_INFO'] =~ /^\/v1\/public\//
      end

      def own_data?(uuid)
        @token_owner['uuid'] == uuid
      end

      def system_level_privileges?
        @token_owner['type'] == 'app'
      end

      def is_authorized_to_access?(uuid)
        system_level_privileges? || own_data?(uuid)
      end

      def prevent_access!
        error!('Unauthenticated', 403)
      end

      def owner_only!(uuid = params[:uuid])
        prevent_access! unless is_authorized_to_access?(uuid)
      end

      def system_privileges_only!
        prevent_access! unless system_level_privileges?
      end

      def venue_identities_for(uuids)
        result = try_twice_and_avoid_token_expiration do
          begin
            if uuids.kind_of?(Array)
              connection.auth.venue_identities_of(token, *uuids)
            else
              connection.auth.venue_identities_of(token, uuids)
            end
          rescue Service::Client::ServiceError => e
            error!(e.error, 404)
          end
        end

        result = Hash[[[uuids, result]]] unless uuids.kind_of?(Array)
        result
      end

      def strip_private_parts_from_venue_identities(venue_identities)
        Hash[venue_identities.map do |venue, venue_identity|
          public_identity = {
            'name' => venue_identity['name']
          }
          [venue, public_identity]
        end]
      end

      def public_venue_identity_info_for(venue_identities_data)
        Hash[venue_identities_data.map do |uuid, venue_identities|
          venue_identities = strip_private_parts_from_venue_identities(venue_identities)
          [uuid, {uuid: uuid, venues: venue_identities}]
        end]
      end

      def get_games_for(uuid, venue = nil)
        games = try_twice_and_avoid_token_expiration do
          uuids = nil
          if venue
            venue = "venue#{Utils.camelize_string(venue)}"
            uuids = connection.graph.query(token, [uuid], "MATCH node0-[p:plays]->game WHERE p.#{venue}! = true RETURN DISTINCT game.uuid").map &:first
          else
            uuids = connection.graph.list_related_entities(uuid, token, 'plays')
          end

          if !uuids || uuids.empty?
            []
          else
            connection.devcenter.list_games(uuids)
          end
        end
        {games: games}
      end

      def empty_body
        {}
      end

      def attach_missing_venue_identity!(raw_venue, player)
        if raw_venue == 'embedded'
          venue_identities = connection.auth.venue_identities_of(token, player)
          if !venue_identities[raw_venue] && !venue_identities.empty?
            existing_venue_identity = venue_identities.values.first
            new_venue_identity = {"venue-id" => player, "name" => existing_venue_identity["name"]}
            connection.auth.attach_venue_identity_to(token, player, raw_venue, new_venue_identity)
          end
        end
      end
    end

    before do
      header('Access-Control-Allow-Origin', request.env['HTTP_ORIGIN'] || '*')

      unless authentication_exception?
        token = request.env['HTTP_AUTHORIZATION'] || params[:oauth_token]
        prevent_access! unless token
        token = token.gsub(/^Bearer\s+/, '')

        @token_owner = connection.auth.token_owner(token)
        prevent_access! unless @token_owner
      end
    end

    options '*path' do
      header('Access-Control-Allow-Headers', 'origin, x-requested-with, content-type, accept, authorization')
      header('Access-Control-Allow-Methods', 'GET, PUT, OPTIONS, POST, DELETE')
      header('Access-Control-Max-Age', '1728000')
      ""
    end

    get "public/players" do
      uuids = params[:uuids]

      venue_identities = venue_identities_for(uuids)
      venue_identities = Hash[[[uuids.first, venue_identities]]] if uuids.length == 1

      public_venue_identity_info_for(venue_identities)
    end

    get "public/:uuid" do
      uuid = params[:uuid]
      venue_identities = venue_identities_for(uuid)
      public_venue_identity_info_for(venue_identities)[uuid]
    end

    get "public/:uuid/friends" do
      requester_uuid = params[:uuid]

      try_twice_and_avoid_token_expiration do
        uuids = connection.graph.list_related_entities(requester_uuid, token, 'friends')

        uuids.unshift requester_uuid
        uuids.uniq!

        identities = connection.auth.venue_identities_of(token, *uuids)

        # Make up for the different response format of the auth-backend
        # depending on if you request venue identities for one or many
        # UUIDs
        identities = {uuids.first => identities} if uuids.size == 1

        Hash[identities.map {|uuid, identities| [uuid, strip_private_parts_from_venue_identities(identities)]}]
      end
    end

    get "/public/:uuid/games" do
      get_games_for(params[:uuid], params[:venue])
    end

    get ":uuid" do
      owner_only!(params[:uuid])

      uuid = params[:uuid]
      venue_identities = venue_identities_for(uuid)
      {uuid: uuid, venues: venue_identities[uuid]}
    end

    # deprecated!
    get ":uuid/games" do
      player = params[:uuid]
      venue = params[:venue]

      last_updated_played_games_at = connection.cache.fetch(['last_game_registered', player, venue]) {-1}

      connection.cache.fetch(['games_i_play', player, venue, last_updated_played_games_at]) do
        get_games_for(player, venue)
      end
    end

    get ":uuid/games/friends" do
      owner_only!(params[:uuid])

      uuid = params[:uuid]
      venue = params[:venue]

      cache_buster_time = Time.now.to_i.div(60 * 60 * 2) # Time to live
      connection.cache.fetch(['games_my_friends_play', uuid, venue, cache_buster_time]) do
        games = try_twice_and_avoid_token_expiration do
          uuids = nil
          if venue
            venue = "venue#{Utils.camelize_string(venue)}"
            uuids = connection.graph.query(token, [uuid], "MATCH node0-[:friends]->()-[p:plays]->game WHERE p.#{venue}! = true RETURN DISTINCT game.uuid")
          else
            uuids = connection.graph.query(token, [uuid], 'MATCH node0-[:friends]->()-[:plays]->game RETURN DISTINCT game.uuid')
          end
          connection.devcenter.list_games(uuids.map &:first)
        end
        {games: games}
      end
    end


    get ":player_uuid/games/:game_uuid/meta-data" do
      owner_only!(params[:player_uuid])

      player = params[:player_uuid]
      game = params[:game_uuid]
      data = try_twice_and_avoid_token_expiration do
        connection.graph.relationship_metadata(player, game, token, 'plays')
      end
      {meta: MetaData.from_graph(data)}
    end


    [":player_uuid/games/:game_uuid/meta-data", ":player_uuid/games/:game_uuid/meta-data/:property"].each do |url|
      put url do
        owner_only!(params[:player_uuid])

        player = params[:player_uuid]
        game = params[:game_uuid]
        data = (params[:meta] || {}).to_hash
        property = params[:property]

        old_data = MetaData.from_graph(connection.graph.relationship_metadata(player, game, token, 'plays'))

        data = try_twice_and_avoid_token_expiration do
          if property
            if data.has_key?(property)
              new_value = data.delete(property)
              new_data = {property => new_value}
              data = old_data.merge(new_data)
            else
              data = old_data
            end
          end

          response = connection.graph.add_relationship(player, game, token, 'plays', meta: MetaData.to_graph(data))
          response.raw.status == 200 ? response.data.first['meta'] : MetaData.to_graph(old_data)
        end
        {meta: MetaData.from_graph(data)}
      end
    end

    post ":player_uuid/games/:game_uuid/:venue" do
      owner_only!(params[:player_uuid])

      player = params[:player_uuid]
      game = params[:game_uuid]
      raw_venue = params[:venue]
      venue = Utils.camelize_string(raw_venue)

      connection.cache.set(['last_game_registered', player, venue], Time.now.to_i)

      connection.cache.fetch(['game_registered_v2', player, venue, game]) do
        try_twice_and_avoid_token_expiration do
          attach_missing_venue_identity!(raw_venue, player)

          connection.graph.add_role(player, token, 'player')
          response = connection.graph.add_relationship(player, game, token, 'plays', meta: {"venue#{venue}" => true})
          status response.raw.status
        end
      end
      empty_body
    end

    get ":uuid/friends" do
      owner_only!(params[:uuid])

      requester_uuid = params[:uuid]
      game = params[:game]
      meta = params[:meta]
      meta = JSON.parse(meta) if meta && meta.kind_of?(String)
      meta_results = nil

      try_twice_and_avoid_token_expiration do
        uuids = nil
        if game
          query = "MATCH node0-[:friends]->friend-[p:plays]->game WHERE game = node1 RETURN DISTINCT friend.uuid"
          query = "#{query}, #{meta.map {|m| "p.#{MetaData::PREFIX}#{m}!"}.join(', ')}" if meta
          uuids = []
          meta_results = {}
          connection.graph.query(token, [requester_uuid, game], query).each do |result|
            uuid = result.shift
            uuids << uuid
            meta_results[uuid] = result
          end
        else
          uuids = connection.graph.list_related_entities(requester_uuid, token, 'friends')
        end
        uuids.unshift requester_uuid
        uuids.uniq!

        identities = connection.auth.venue_identities_of(token, *uuids)

        # Make up for the different response format of the auth-backend
        # depending on if you request venue identities for one or many
        # UUIDs
        identities = {uuids.first => identities} if uuids.size == 1

        if meta && meta_results
          requester_meta_query  = "MATCH node0-[p:plays]->game WHERE game = node1 RETURN DISTINCT node0.uuid"
          requester_meta_query  = "#{requester_meta_query }, #{meta.map {|m| "p.#{MetaData::PREFIX}#{m}!"}.join(', ')}"
          result = connection.graph.query(token, [requester_uuid, game], requester_meta_query ).first
          meta_results[requester_uuid] = result[1..-1]

          identities = Hash[identities.map {|uuid, identity| [uuid, identity.merge('meta' => Hash[meta.zip(meta_results[uuid])])]}]
        end

        identities
      end
    end

    put ":uuid/friends/:venue_id" do
      owner_only!(params[:uuid])

      friends_data = params[:friends]

      venue_id = params[:venue_id]

      cache_buster_time = Time.now
      connection.cache.fetch(['update_friends', 'v1', params[:uuid], params[:venue_id], cache_buster_time.year, cache_buster_time.month, cache_buster_time.day]) do
        Futuroscope::Future.new do
          connection = Connection.create
          friend_uuids = try_twice_and_avoid_token_expiration do
            connection.auth.uuids_of(token, venue_id => friends_data)[venue_id]
          end

          venue = Venue.const_get(Utils.camelize_string(venue_id)).new
          friend_uuids.values.each do |friend_uuid|
            try_twice_and_avoid_token_expiration do
              connection.graph.add_role(friend_uuid, token, 'player')
              venue.friend(params[:uuid], friend_uuid, token, connection)
            end
          end
        end
      end

      empty_body
    end

    get ":uuid/avatars/:venue_id" do
      venue_identities = try_twice_and_avoid_token_expiration do
        connection.auth.venue_identities_of(token, params[:uuid])
      end
      # identity = venue_identities[params[:venue_id]]
      identity = venue_identities['facebook'] # <-- Dirty hack till we have real friends on all venues

      redirect "https://graph.facebook.com/#{identity['id']}/picture"
    end
  end
end

